native/src/main/java/io/ballerina/stdlib/http/transport/contractimpl/listener/states/SendingEntityBody.java

```diff
@@ -270,9 +289,66 @@ private void triggerPipeliningLogic(HttpCarbonMessage outboundResponseMsg) {
         }
     }
 
+    private void logAccessInfo(HttpCarbonMessage inboundRequestMsg, HttpCarbonMessage outboundResponseMsg) {
+        HttpHeaders headers = inboundRequestMsg.getHeaders();
+        if (headers.contains(HTTP_X_FORWARDED_FOR)) {
+            String forwardedHops = headers.get(HTTP_X_FORWARDED_FOR);
+            // If multiple IPs available, the first ip is the client
+            int firstCommaIndex = forwardedHops.indexOf(',');
+            remoteAddress = firstCommaIndex != -1 ? forwardedHops.substring(0, firstCommaIndex) : forwardedHops;
+        }
+
+        // Populate request parameters
+        String userAgent = "-";
+        if (headers.contains(HttpHeaderNames.USER_AGENT)) {
+            userAgent = headers.get(HttpHeaderNames.USER_AGENT);
+        }
+        String referrer = "-";
+        if (headers.contains(HttpHeaderNames.REFERER)) {
+            referrer = headers.get(HttpHeaderNames.REFERER);
+        }
+        String method = inboundRequestMsg.getHttpMethod();
+        String uri = inboundRequestMsg.getRequestUrl();
+        HttpMessage request = inboundRequestMsg.getNettyHttpRequest();
+        String protocol;
+        if (request != null) {
+            protocol = request.protocolVersion().toString();
+        } else {
+            protocol = inboundRequestMsg.getHttpVersion();
+        }
+
+        // Populate response parameters
+        int statusCode = Util.getHttpResponseStatus(outboundResponseMsg).code();
+
+        long requestTime = Calendar.getInstance().getTimeInMillis() - inboundRequestArrivalTime.getTimeInMillis();
+        HttpAccessLogMessage inboundMessage = new HttpAccessLogMessage(remoteAddress,
+                inboundRequestArrivalTime, method, uri, protocol, statusCode, contentLength, referrer, userAgent);
+        inboundMessage.setRequestBodySize((long) inboundRequestMsg.getContentSize());
+        inboundMessage.setRequestTime(requestTime);
+
+        List<HttpAccessLogMessage> outboundMessages = getHttpAccessLogMessages(inboundRequestMsg);
+
+        HttpAccessLogger.log(inboundMessage, outboundMessages);
+    }
+
     private void resetOutboundListenerState() {
         contentList.clear();
         contentLength = 0;
         headersWritten = false;
     }
+
+    private List<HttpAccessLogMessage> getHttpAccessLogMessages(HttpCarbonMessage request) {
+        Object outboundAccessLogMessagesObject = request.getProperty(OUTBOUND_ACCESS_LOG_MESSAGES);
+        if (outboundAccessLogMessagesObject instanceof List<?> rawList) {
+            for (Object item : rawList) {
```

Comment:
is there a possibility for the rawList to contain other LogMessage types than `HttpAccessLogMessage`?

Reply:
ATM, no only `HttpAccessLogMessage`